astropy/stats/funcs.py

```diff
@@ -1674,12 +1674,11 @@ def histogram_intervals(
 
     """
     h = np.zeros(n)
-    start = breaks[0]
-    for i in range(len(totals)):
-        end = breaks[i + 1]
-        for j in range(n):
-            ol = interval_overlap_length((float(j) / n, float(j + 1) / n), (start, end))
-            h[j] += ol / (1.0 / n) * totals[i]
+    start = n * breaks[0]
+    for i, total in enumerate(totals):
+        end = n * breaks[i + 1]
+        for j in range(max(0, math.floor(start)), min(n, math.ceil(end))):
```

Comment:
For valid `breaks` this could be written as
```suggestion
        for j in range(math.floor(start), math.ceil(end)):
```
but if `breaks` contains values outside $[0, 1]$ then that would cause an `IndexError` whereas the code on current `main` handles such invalid input silently. The more complicated code is therefore required to avoid an API change.

Reply:
Feel free to submit a separate PR for these changes.